### DISCUSSION TITLE
[2팀: U'STOCK] BE 운영경험

### DISCUSSION THREAD

**Discussion by Lcrescent3es** [100-hours-a-week/discussions#101] [Story]
### 1. **트래픽 급증 시 서비스가 어떻게 대응했나요?**
- 실제로 트래픽이 몰린 적은 없지만, 로드 테스트를 진행했을 때, 응답 시간이 길어지는 현상을 확인했습니다. DB INSERT 테스트에서 평상시 30~50ms의 응답 속도를 보여주던 데 반해 스레드 100개가 지속적으로 요청하는 상황에서 최대 360ms, 스레드 1개당 평균 3ms의 지연이 발생하였습니다. 따로 해결을 하진 않았지만, 서버 수를 늘린 뒤 로드 밸런싱과 READ-ONLY DB 세팅으로 지연을 줄일 수 있을 것 같습니다.

### 2. **서비스 장애나 오류가 사용자에게 어떻게 노출되었나요?**
- 백엔드에서 서비스 장애나 오류가 발생할 시 예외처리를 거쳐 FE에 응답을 보내주면 FE에서 사용자에게 대응을 해야하는데, 기존 서비스의 BE 예외처리 방식은 해당 코드에서 예외를 처리하여 응답 양식을 맞추기 힘들었고, unhandled exception으로 인해 spring 서버가 평균 2일 중 1번씩 다운되었습니다.  해당 문제를 해결하기 위해 전역 예외 핸들러로 예외 처리 방식을 개선하였고, 장애 발생시 custom exception으로 message를 전달한 후 fe측에서 해당 오류 메시지 내용에 맞춰 500 error는 장애 페이지 제공, 외에는 사용자에게 모달창으로 제공했습니다.

### 3. **다운타임 없이 유지보수를 진행한 경험이 있나요?**
- devops 담당 인원이 aws ec2 인스턴스에 Github Action과 docker container를 이용하여 Blue-Green 무중단 배포를 도입하였고, 다운타임 없이 서비스 업데이트를 진행하였습니다.
배포 도중, port가 바뀌고 container가 중지되는 과정에서 요청이 들어왔을 때, 해당 요청으로 인해 장애가 발생하지만, 장애 페이지를 제공하여 사용자에게 장애 사실을 알렸습니다.

### 4. **사용자 피드백을 반영해 성능을 개선한 경험이 있나요?**
- QA중 특정 요청이 응답까지 오랜 시간이 걸린다는 피드백을 받았습니다. 확인 결과 1000ms 이상이 소요되었고, JPA의 lazy loading으로 인한 DB 연결 오버헤드 때문이었습니다. 팀원과 함께 JPQL로 직접 쿼리를 작성해 해당 문제를 해결하였고, 30~50ms이내로 응답 시간을 단축하였습니다.

### 5. **실제 서비스 운영 중 사용자 데이터 유실 문제를 어떻게 해결했나요?**
- DB로 mysql(RDB)과 redis(Elasticache)를 사용했는데, mysql의 백업 기능과 redis의 sentinel, master-slave, 클러스터링 아키텍쳐를 도입한다면 해결될 문제인 것 같습니다. 

### 6. **사용자 로그나 통계를 어떻게 관리하고 분석했나요?**
- FE측에서 Google Analytics 와 Microsoft Clarity를 도입하여 사용자 로그와 행동 패턴을 수집하였고, BE에선 로그와 sentry를 통해 간접적으로 장애가 발생한 위치를 파악하였습니다.
sentry를 통해 이슈 트래킹을 진행하여 미리 감지하지 못한 20가지 정도의 여러 이슈들을 심각도별로 분류하여 해결하였고, 서비스의 내구성을 향상시켰습니다. 

### 7. **실사용자에게 직접적으로 영향을 미친 가장 큰 이슈는 무엇이었나요?**
- 로그인 시 리프레시 토큰을 검증하는 로직에 문제가 있어서, 유효하지 않은 리프레시 토큰으로 인해 사용자의 로그인 시간이 길어지면 기능을 제대로 이용할 수 없는 문제가 있었습니다. 해당 이슈로 인해 복귀한 사용자가 서비스를 이용하지 못해 이탈하는 상황이 있었고, 피드백 폼과 QA, sentry를 통해 이슈를 트래킹하여 처리하였습니다.

### 8. **서비스 운영 중 사용자 불만을 처리한 경험이 있나요?**
- 게임 실행 시 연도가 넘어갈 때 마다 속도가 너무 느리다는 불만이 있었는데, 확인결과 JPA OneToOne 관계에서의 Eager Loading과 AI 응답 시간이 문제였습니다. 이후 팀원과 함께 JPA 쿼리 최적화를 통해 최적화를 진행하였고, ai 요청을 제외한 응답 시간을 3분의 1로 줄였습니다.

### 9. **사용자 피드백을 받았을 때 이를 시스템 개선에 어떻게 반영했나요?**
-

**Comment by ej31** [comment]
공성전 한번 했어야 하는데.. 스읍..
